### ISSUE TITLE
Re-investigate fire-related systems

### ISSUE THREAD

**Issue by gtfierro** [BrickSchema/Brick#556]
There is some ambiguity in our classes around fire control systems. According to https://www.nfpa.org/News-and-Research/Publications-and-media/Blogs-Landing-Page/NFPA-Today/Blog-Posts/2021/03/03/A-Guide-to-Fire-Alarm-Basics we may want to have 3 broad types of equipment: fire control panel, fire alarm initiation systems, and fire alarm notification systems.

**Comment by jbkoh** [comment]
Would like to add suppressing systems like extinguishers and sprinklers.

**Comment by jbkoh** [comment]
Also wonder if detectors could be an alarm at the same time in many cases (like home fire alarms)

**Comment by gtfierro** [comment]
> Also wonder if detectors could be an alarm at the same time in many cases (like home fire alarms)

We can probably create an Alarm Point that we associate with the detectors to draw a distinction between the device and the signal

Or are you asking if it is both the "detector" and the "sound emitter"?

**Comment by jbkoh** [comment]
> Or are you asking if it is both the "detector" and the "sound emitter"?
Yes. Those detectors emit sound as well, so either we have to adjust the hierarchy or let users instantiate both at the same time. I 'guess' it's common to have both functions.

**Comment by gtfierro** [comment]
According to the conversation I'm having on the REC gitter, future versions of DTDL (v3 in particular) will support up to 1024-way multiple inheritance. Azure Digital Twins v3 is still in preview I guess, so we don't know if REC will migrate to DTDLv3 yet. If they do, it will probably be after ADTv3 moves out of preview

**Comment by ektrah** [comment]
> > Or are you asking if it is both the "detector" and the "sound emitter"?
> 
> Yes. Those detectors emit sound as well, so either we have to adjust the hierarchy or let users instantiate both at the same time. I 'guess' it's common to have both functions.

Yes, I think it's quite common to have devices that combine multiple aspects (e.g., a point detector that detects both heat and smoke).

A sound-emitting fire detector, I think, I would model as a combination of a `Fire_Detector` and an `Audible_Alarm_Device`.

> According to the conversation I'm having on the REC gitter, future versions of DTDL (v3 in particular) will support up to 1024-way multiple inheritance. Azure Digital Twins v3 is still in preview I guess, so we don't know if REC will migrate to DTDLv3 yet. If they do, it will probably be after ADTv3 moves out of preview

Great, then perhaps we can avoid the combinatorial explosion by using multiple inheritance here.

What would be the most <s>pythonic</s> [<s>brickish</s>](https://en.wiktionary.org/wiki/brickish) [bricky](https://en.wiktionary.org/wiki/bricky) way of doing this, e.g., for the `Fire_Detector`? Should there be one subclass for each item along each dimension?

- flame detectors, smoke detectors, heat detectors, carbon monoxide detectors;
- point detectors, line type detectors, multipoint detectors, aspirating detectors;
- resettable detectors, non-resettable detectors;
- detachable detectors, non detachable detectors.

**Comment by ektrah** [comment]
Relevant past discussion: https://github.com/BrickSchema/Brick/pull/336

**Comment by ektrah** [comment]
@gtfierro @jbkoh @epaulson I would like to ask for quick feedback as to whether the proposed terms in https://github.com/BrickSchema/Brick/issues/556#issuecomment-1708900836 are OK. If so, I would prepare a pull request so that the new equipment types can become part of Brick 1.4.

**Comment by jbkoh** [comment]
Though I like the hierarchy, I'm not sure if we have an agreement on the audible fire detector issue. Is the solution to have "Fire_Detector_with_Audible_Alarm"?

**Comment by jbkoh** [comment]
Btw, I'm looking to define Smoke Zones and Fire Zones too: https://ltc.health.mo.gov/archives/15059#:~:text=The%20fire%20alarm%20zones%20are,full%20evacuation%20of%20the%20building.
Maybe this should go to REC though.

**Comment by connorjcantrell** [comment]
@ektrah @jbkoh I suggest changing the term "Carbon Monoxide Fire Detector" to simply "Carbon Monoxide Detector." These devices primarily serve to identify the presence of CO gas originating from malfunctioning boilers, water heaters, leaking exhaust vents, or vehicles in enclosed parking areas. While they are associated with combustible sources, their primary function is not directly connected to fire detection or fire safety.

**Comment by ektrah** [comment]
As I understand it, on the one hand there are detectors that have the purpose of detecting fire and of which there are a number of variants that differ in the detection method (e.g., based on the presence of smoke, heat, flames, carbon monoxide). And on the other hand, there are detectors whose purpose is to detect the presence of gases. I'm not sure if carbon monoxide fire detectors and carbon monoxide detectors are interchangeable enough that we should combine them, even though they probably work internally very similarly. What do you think about having both?

**Comment by jbkoh** [comment]
Hi @ektrah yeah, we can have both and let the fire one be a subclass of the
other one. However, if we distinguish "alarm device" from "detectors", sole
detectors are equivalent to just "sensor equipment", and in that case
having a separate class of "CO Detector" might be not super necessary.

So, I have to ask the question that I raised again; are "detectors"
dedicated to just sensing something alone if we have Audible_Alarm_Device
separately? In that case, what are the distinguishable features for
detectors compared to Sensor Equipment's?

With regards,
Jason Koh

**Comment by ektrah** [comment]
Here is how I understand it:

- A **sensor**&sup1; is a device that measures a physical quantity (such as temperature, humidity, flowrate, etc.).
- An **indoor weather station** is a box that has the purpose of measuring the temperature and humidity in a room and has points that reflect these measurements. Internally, an indoor weather station might have a number of sensors&sup1; that perform the actual measurements, but these parts are usually not explicitly modeled as we are more interest in the points of the box.
- A **water meter** is a box that has the purpose of metering the volume of water consumed since some starting point and has a point that reflects these meterings. Internally, a water meter might have a sensor&sup1; that measures the flowrate of water, but this part is usually not modeled and there is no point that reflects these flowrate measurements.
- A **fire detector** is a box that detects the presence of a fire and has a point that reflects these detections. Internally, a fire detector might have a sensor&sup1; that measures the concentration of carbon monoxide in the air surrounding the box, but this part is usually not modeled and there is no point that reflects these concentration measurements. It's useful to indicate the detection method, though.
- A **carbon monoxide detector** is a box that has the purpose of measuring the concentration of carbon monoxide in the air surrounding the box. Internally, a carbon monoxide detector might have a sensos&sup1; that performs the actual measurements, but this parts are usually not explicitly modeled as we are more interest in the points of the box.
- In general, the different kinds of fire detectors and manual alarm initiating devices have the purpose of signaling a fire to a **fire alarm control panel**. The fire alarm control panel has the purpose of deciding how to react (sound an alarm, call the fire brigade, close fire dampers, ...) and sending commands to devices accordingly.
- An **audible alarm device** is a box that has the purpose of making noise to alert the occupants of a fire and that has a point for receiving commands.

The open question from my point of view is how to model a box that combines a fire detector and an audible alarm device.

(&sup1; in the sense of `s223:Sensor`, not `brick:Sensor` or `brick:Sensor_Equipment`)
